diff -Nuar oregoncore-current/src/game/Player.cpp oregoncore-pq/src/game/Player.cpp
--- oregoncore-current/src/game/Player.cpp	2011-01-01 10:03:08.000000000 -0500
+++ oregoncore-pq/src/game/Player.cpp	2011-01-01 10:11:22.000000000 -0500
@@ -857,7 +857,11 @@
     switch (timer)
     {
         case FATIGUE_TIMER:
+        {
+            if (!isAlive() || MirrorTimerType(FATIGUE_TIMER) ||GetSession()->GetSecurity() >= sWorld.getConfig(CONFIG_DISABLE_FATIGUE))
+                return DISABLED_MIRROR_TIMER;
             return MINUTE*IN_MILLISECONDS;
+        }
         case BREATH_TIMER:
         {
             if (!isAlive() || HasAuraType(SPELL_AURA_WATER_BREATHING) || GetSession()->GetSecurity() >= sWorld.getConfig(CONFIG_DISABLE_BREATHING))
diff -Nuar oregoncore-current/src/game/World.cpp oregoncore-pq/src/game/World.cpp
--- oregoncore-current/src/game/World.cpp	2011-01-01 10:03:08.000000000 -0500
+++ oregoncore-pq/src/game/World.cpp	2011-01-01 10:11:22.000000000 -0500
@@ -892,6 +892,7 @@
     m_configs[CONFIG_WEATHER] = sConfig.GetBoolDefault("ActivateWeather",true);
 
     m_configs[CONFIG_DISABLE_BREATHING] = sConfig.GetIntDefault("DisableWaterBreath", SEC_CONSOLE);
+    m_configs[CONFIG_DISABLE_FATIGUE] = sConfig.GetIntDefault("DisableFatigue", SEC_CONSOLE);
 
     m_configs[CONFIG_ALWAYS_MAX_SKILL_FOR_LEVEL] = sConfig.GetBoolDefault("AlwaysMaxSkillForLevel", false);
 
diff -Nuar oregoncore-current/src/game/World.h oregoncore-pq/src/game/World.h
--- oregoncore-current/src/game/World.h	2011-01-01 10:03:08.000000000 -0500
+++ oregoncore-pq/src/game/World.h	2011-01-01 10:11:22.000000000 -0500
@@ -195,6 +195,7 @@
     CONFIG_CHANCE_OF_GM_SURVEY,
     CONFIG_INSTANT_LOGOUT,
     CONFIG_DISABLE_BREATHING,
+    CONFIG_DISABLE_FATIGUE,
     CONFIG_ALL_TAXI_PATHS,
     CONFIG_DECLINED_NAMES_USED,
     CONFIG_LISTEN_RANGE_SAY,
diff -Nuar oregoncore-current/src/oregoncore/oregoncore.conf.dist oregoncore-pq/src/oregoncore/oregoncore.conf.dist
--- oregoncore-current/src/oregoncore/oregoncore.conf.dist	2011-01-01 10:03:08.000000000 -0500
+++ oregoncore-pq/src/oregoncore/oregoncore.conf.dist	2011-01-01 10:11:22.000000000 -0500
@@ -692,6 +692,10 @@
 #        Disable/enable waterbreathing for security level (0..4) or higher
 #        Default: 4 (None)
 #
+#    DisableFatigue
+#        Disable/enable fatigue for security level (0..4) or high
+#        Default: 4 (None)
+#
 #    AllFlightPaths
 #        Players will start with all flight paths
 #         (Note: ALL flight paths, not only player's team)
@@ -824,6 +828,7 @@
 StartArenaPoints = 0
 InstantLogout = 1
 DisableWaterBreath = 4
+DisableFatigue = 4
 AllFlightPaths = 0
 AlwaysMaxSkillForLevel = 0
 ActivateWeather = 1
